Decouple memory allocations from aligned memory allocations

npitre opened the pull request:
When a simple memory allocation is accomplished in terms of an aligned memory
allocation, a longer aligned allocation code path with an extra runtime
overhead is involved even though no alignment is necessary.

Let's rework the code so the aligned allocation code path is invoked only
when an actual aligned allocation is requested. This opens the possibility
for the linker to garbage-collect the aligning code otherwise.

npitre committed:
kernel: k_aligned_alloc: simplify assertion

Let's avoid division and modulus operations as they're costly... and even
more so when they're unnecessary as in this case the main constraint is
about the alignment being a power of 2 which is a very small subset of
sizeof(void *) multiples.

Then move the assertion to common code for wider coverage.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
kernel: kheap: decouple simple alloc from aligned_alloc

When k_heap_alloc() is expressed in terms of k_heap_aligned_alloc()
it invokes a longer aligned allocation code path with an extra runtime
overhead even though no alignment is necessary.

Let's reference and invoke the aligned allocation code path only when an
actual aligned allocation is requested. This opens the possibility for
the linker to garbage-collect the aligning code otherwise.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
kernel: mempool: decouple simple alloc from aligned_alloc

When k_malloc() is expressed in terms of k_aligned_alloc() it invokes a
longer aligned allocation code path with an extra runtime overhead even
though no alignment is necessary.

Let's reference and invoke the aligned allocation code path only when an
actual aligned allocation is requested. This opens the possibility for
the linker to garbage-collect the aligning code otherwise.

Also bypass k_heap_malloc() and friends given they're invoked with
K_NO_WAIT. Go directly to sys_heap_*() instead to cut some more unneeded
overhead.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
kernel: sys_heap: decouple realloc from aligned_realloc

When sys_heap_realloc() is expressed in terms of sys_heap_aligned_realloc()
it invokes a longer aligned allocation code path with an extra runtime
overhead even though no alignment is necessary.

Let's reference and invoke the aligned allocation code path only when an
actual aligned allocation is requested. This opens the possibility for
the linker to garbage-collect the aligning code otherwise.

Improve realloc documentation while at it.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

peter-mitsis reviewed (commented):
Just one question/comment about alignment in z_heap_aligned_alloc(). Everything else looks good to me.

npitre commented on kernel/mempool.c:
```diff
@@ -15,6 +15,9 @@ static void *z_heap_aligned_alloc(struct k_heap *heap, size_t align, size_t size
 	struct k_heap **heap_ref;
 	size_t __align;
 
+	__ASSERT((align & (align - 1)) == 0,
+		"align must be a power of 2");
```
Minimum alignment from heap.c is 4 bytes on 32-bits archs and 8 bytes on 64-bits archs already.

andyross reviewed (dismissed):
Seems very reasonable.  Doesn't look like the impact on code size should be very high, but might be worth measuring.

cfriedt commented on lib/heap/heap.c:
```diff
@@ -459,22 +457,66 @@ void *sys_heap_aligned_realloc(struct sys_heap *heap, void *ptr,
+		return NULL;
+	}
+
+	__ASSERT((align & (align - 1)) == 0, "align must be a power of 2");
```
Just in case there is another respin
```suggestion
	__ASSERT(IS_POWER_OF_TWO(align));
```

cfriedt replied:
Ah, align of 0 would trigger the assert as well and this macro would not work in that case.

cfriedt reviewed (dismissed):
Looks good to me. Just one suggestion, but not a blocker (use `IS_POWER_OF_TWO()`).

Re: measurements, I was curious about the size difference and as well, and also a little curious about timing. If there are fewer conditions on a potentially hot path like this, there could be some perf gain.

Would be really great to see some memory allocation benchmarks down the road.

npitre commented:
Due to popular demand, here's some code size comparisons using the 
qemu_cortex_a53 target.

tests/lib/heap_align:

```
   text    data     bss     dec     hex filename
  60928    4800  339966  405694   630be zephyr_before.elf
  60928    4800  339966  405694   630be zephyr_after.elf
```

tests/kernel/mem_heap/k_heap_api:

```
   text    data     bss     dec     hex filename
  61113    4704  344066  409883   6411b zephyr_before.elf
  61057    4760  344066  409883   6411b zephyr_after.elf
```

samples/userspace/hello_world_user:

```
   text    data     bss     dec     hex filename
  75478   24716  356781  456975   6f90f zephyr_before.elf
  75478   24716  356781  456975   6f90f zephyr_after.elf
```

So... it seems there isn't much difference as code size goes. If anything, 
the code may be slightly smaller, but the data section grows accordingly
probably due to additional alignment padding.

TaiJuWu commented on kernel/kheap.c:
```diff
@@ -110,13 +109,27 @@ void *k_heap_alloc(struct k_heap *heap, size_t bytes, k_timeout_t timeout)
 	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_heap, alloc, heap, timeout);
 
-	void *ret = k_heap_aligned_alloc(heap, sizeof(void *), bytes, timeout);
+	void *ret = z_heap_alloc_helper(heap, 0, bytes, timeout,
```
Should we add comments `0 is ignored`?

npitre replied:
Documentation for `sys_heap_noalign_alloc()` mentions this argument is ignored.

npitre committed:
tests: benchmarks: sys_kernel: add k_malloc() test

Useful to evaluate malloc performance changes.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

cfriedt commented on kernel/mempool.c:
```diff
@@ -9,11 +9,18 @@
 	size_t __align;
+	k_spinlock_key_t key;
+
+	__ASSERT((align & (align - 1)) == 0,
+		"align must be a power of 2");
```
I still think that a descriptive macro would be better here.
```suggestion
	__ASSERT(IS_POWER_OF_TWO(align));
```

npitre replied:
Can't use that as 0 must be a valid value too.

npitre committed:
kernel: kheap: workaround for Percepio's TraceRecorder kludge

In .../modules/debug/percepio/TraceRecorder/kernelports/Zephyr/include/\
tracing_tracerecorder.h there is a concealed non-parameterized direct
reference to a local variable that is no longer in scope. Provide a dummy
stub for compilation to succeed until that module's layering violation is
fixed, after which this could be reverted.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
And here's some perf numbers using qemu_riscv64 which I hope can be
meaningful enough. That's from tests/benchmarks/sys_kernel with the first
commit applied in both cases.

Before malloc changes:

```
TEST CASE: malloc #1
TEST COVERAGE:
        k_malloc
Starting test. Please wait...
TEST RESULT: SUCCESSFUL
DETAILS: Average time for 1 iteration: 29890 nSec
END TEST CASE

TEST CASE: malloc #2
TEST COVERAGE:
        k_free
Starting test. Please wait...
TEST RESULT: SUCCESSFUL
DETAILS: Average time for 1 iteration: 22640 nSec
END TEST CASE
```

With the whole series applied:

```
TEST CASE: malloc #1
TEST COVERAGE:
        k_malloc
Starting test. Please wait...
TEST RESULT: SUCCESSFUL
DETAILS: Average time for 1 iteration: 25920 nSec
END TEST CASE

TEST CASE: malloc #2
TEST COVERAGE:
        k_free
Starting test. Please wait...
TEST RESULT: SUCCESSFUL
DETAILS: Average time for 1 iteration: 22640 nSec
END TEST CASE
```

In short, `k_malloc()` performance improved by 13%.

npitre commented:
@andyross, @peter-mitsis: ping

npitre commented:
This is no April Fools' prank: assignees' (re)approval is still needed. ;-)
